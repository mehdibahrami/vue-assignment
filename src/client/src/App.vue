<template>
  <Header />
  <Footer />
</template>

<script>
import { Header, Footer } from "@/layouts";

export default {
  name: "App",
  components: {
    Header,
    Footer,
  },
  data() {
    return {
      vehicleStatus: {
        time: 0,
        energy: 0,
        gps: {
          lat: 0,
          lng: 0,
        },
        odo: 0,
        speed: 0,
        soc: 0,
      },
    };
  },
  created() {
    // Connect to WebSocket
    this.connectToWebsocket();
  },
  methods: {
    connectToWebsocket() {
      // Create a new Webscoket
      const connection = new WebSocket(process.env.VUE_APP_WEBSOCKET_URL);
      // WebSocket is open
      connection.onopen = () => {
        console.log("WebSocket is open now.");
      };
      // WebSocket message received
      connection.onmessage = ({ data }) => {
        this.websocketDataParser(data);
      };
      // Error event occurs on the WebSocket
      connection.onerror = (error) => {
        console.log("WebSocket error observed:", error);
      };
    },
    websocketDataParser(data) {
      // Parse the WebSocket data
      const jsonData = JSON.parse(data);
      // Split the GPS data to an array
      const location = jsonData.gps.split("|");
      // Store the parsed data
      this.vehicleStatus = {
        time: parseInt(jsonData.time),
        energy: parseFloat(jsonData.energy),
        gps: {
          lat: parseFloat(location[0]),
          lng: parseFloat(location[1]),
        },
        odo: parseFloat(jsonData.odo),
        speed: parseInt(jsonData.speed),
        soc: parseFloat(jsonData.soc),
      };
    },
  },
};
</script>
