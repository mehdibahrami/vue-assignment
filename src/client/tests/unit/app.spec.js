import { shallowMount } from "@vue/test-utils";
import App from "@/App.vue";

describe("App", () => {
  let wrapper;

  beforeEach(() => {
    // set env values
    process.env = {
      VUE_APP_WEBSOCKET_URL: "ws://localhost:3000",
    };
  });
  afterAll((done) => {
    done();
  });

  it("renders correctly", () => {
    // shallow mounting the main component
    wrapper = shallowMount(App);

    expect(wrapper.findComponent({ name: "Header" }).exists()).toBeTruthy();
    expect(wrapper.findComponent({ name: "Dashboard" }).exists()).toBeTruthy();
    expect(wrapper.findComponent({ name: "Footer" }).exists()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("connects to websocket", () => {
    // mocking websocket using jest
    const connectToWebsocket = jest.spyOn(App.methods, ["connectToWebsocket"]);
    // shallow mounting the main component
    shallowMount(App);

    expect(connectToWebsocket).toHaveBeenCalledTimes(1);
  });
});
