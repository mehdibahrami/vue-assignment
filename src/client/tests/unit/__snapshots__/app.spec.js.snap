// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <dashboard-stub
    data="[object Object]"
  />
  <footer-stub />
  
</div>
`;
